lib/MatrixGitterBridge.js

```diff
@@ -0,0 +1,302 @@
+      },
+
+      onEvent: function(req, context) {
+        var event = req.getData();
+
+        if (event.type == "m.room.member" &&
+            event.state_key == bridge.getBot().getUserId()) {
+          var membership = event.content.membership;
+          if (membership == "invite") {
+            self.onBotInvited(event.room_id);
+          }
+          else if (membership == "join") {
+            self.onBotJoined(event.room_id);
+          }
+          else if (membership == "leave") {
+            // TODO: "leave" events might mean we left the room, or got kicked
+            //   while still in invite state - i.e. invite was cancelled
+            //   before we joined.
+            self.onBotLeft(event.room_id);
+          }
+          else {
+            console.log('matrix member event for myself to state ' + membership + 
+                        ' from state ', event);
+          }
+          return;
+        }
+
+        if (event.type !== "m.room.message" || !event.content) {
+          return;
+        }
+
+        if (event.sender == bridge.getBot().getUserId()) {
+          return;
+        }
+
+        console.log('matrix->' + event.room_id + ' from ' + event.user_id + ':', event.content.body)
+
+        var handled = false;
+
+        if (config.matrix_admin_room && event.room_id == config.matrix_admin_room) {
+          self.onMatrixAdminMessage(event);
+          handled = true;
+        }
+
+        var bridgedRoom = self._bridgedRoomsByMatrixId[event.room_id];
+        if (bridgedRoom) {
+          bridgedRoom.onMatrixMessage(event);
+          handled = true;
+        }
+
+        if (!handled) {
+          console.log("  Wasn't expecting this room; ignore");
+        }
+      }
+    }
+  });
+
+  this._bridge = bridge;
+  this._gitterUserId = null;
+
+  this._gitterusers = {};
+
+  var rules = [];
+  if (config.name_mangling) {
+    for (var i = 0; i < config.name_mangling.length; i++) {
+      var rule = config.name_mangling[i];
+
+      rules.push({
+        pattern: new RegExp(rule.pattern),
+        template: rule.template
+      });
+    }
+  }
+  this._name_mangling_rules = rules;
+}
+
+MatrixGitterBridge.prototype.getMyGitterUserId = function() {
+  if (this._gitterUserId)
+    return Promise.resolve(this._gitterUserId);
+
+  return this._gitter.currentUser().then((u) => {
+    this._gitterUserId = u.id;
+    return u.id;
+  });
+};
+
+MatrixGitterBridge.prototype.mangleName = function(name) {
+  var rules = this._name_mangling_rules;
+
+  for (var i = 0; i < rules.length; i++) {
+    var rule = rules[i];
+    var matches = rule.pattern.exec(name);
+    if (!matches)
+      continue;
+
+    // TODO: more groups?
+    return rule.template.replace('$1', matches[1]);
+  }
+
+  return name;
+}
+
+MatrixGitterBridge.prototype.mapGitterUser = function(user) {
+  var id = user.id;
+
+  return this._gitterusers[id] = (this._gitterusers[id] ||
+      this.getRemoteUser(id).then((model) => {
```

Comment:
Well excuse me for thinking that `this._gitterusers` was a map of `user_id => GitterUser`. It is in fact a map of `user_id => Promise<GitterUser, Error>`. This is confusing. Please don't do this. Especially when you start plucking things out the map and call them `user` - https://github.com/matrix-org/matrix-appservice-gitter/pull/8/files#diff-73f5f6969c2293ed1df0158153678a6aR150 - someone may actually think it is a `GitterUser`.

Don't get me wrong, I have no problem with you storing Promises (especially since they won't "expire" as such), my beef is with the name.